BasePath parameters are referenced by a Parent function causes a panic

Compiling a design whose `BasePath` parameters are referenced by a `Parent` function causes a panic.

This panic seems to occur only in goa 1.2.

## Example design

```go
package design

import (
	. "github.com/goadesign/goa/design"
	. "github.com/goadesign/goa/design/apidsl"
)

var _ = API("panic", func() {
})

var _ = Resource("parent", func() {
	BasePath("/:SomeParam")
	CanonicalActionName("index")

	Action("index", func() {
		Routing(GET("/"))
		Response(OK)
	})
})

var _ = Resource("panic", func() {
	Parent("parent")
	BasePath("panic")

	Action("panic", func() {
		Routing(GET("/"))
		Response(OK)
	})
})
```

**Log**
```
$ goagen bootstrap -d sample/design
exit status 2
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x12fe7cb]

goroutine 1 [running]:
github.com/goadesign/goa/design.(*dupper).DupType(0xc420201510, 0x1648be0, 0xc4201bff20, 0xc420201580, 0xc420201520)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/dup.go:67 +0x1db
github.com/goadesign/goa/design.Dup(0x1648be0, 0xc4201bff20, 0xc420201870, 0xc420201950)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/dup.go:7 +0x15d
github.com/goadesign/goa/design.(*ActionDefinition).initQueryParams(0xc420093900)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/definitions.go:1610 +0x2d1
github.com/goadesign/goa/design.(*ActionDefinition).Finalize(0xc420093900)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/definitions.go:1480 +0x98
github.com/goadesign/goa/design.(*ResourceDefinition).Finalize.func2(0xc420093900, 0xc4201bfad0, 0x143bc2c)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/definitions.go:875 +0x3d
github.com/goadesign/goa/design.(*ResourceDefinition).IterateActions(0xc4201c60e0, 0xc420201c90, 0x0, 0x0)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/definitions.go:706 +0x1d2
github.com/goadesign/goa/design.(*ResourceDefinition).Finalize(0xc4201c60e0)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/definitions.go:877 +0x138
github.com/goadesign/goa/dslengine.finalizeSet(0xc4201c8460, 0x2, 0x2, 0x0, 0x2)
        /Users/ikawaha/go/src/github.com/goadesign/goa/dslengine/runner.go:330 +0x8e
github.com/goadesign/goa/design.(*APIDefinition).IterateSets(0xc42008fa20, 0x146aec8)
        /Users/ikawaha/go/src/github.com/goadesign/goa/design/definitions.go:498 +0x569
github.com/goadesign/goa/dslengine.Run(0x16415a0, 0xc4200d7da0)
        /Users/ikawaha/go/src/github.com/goadesign/goa/dslengine/runner.go:114 +0x20b
main.main()
        /Users/ikawaha/go/src/cellar/goagen831393131/main.go:25 +0x5a
```

--- a/design/definitions.go
+++ b/design/definitions.go
@@ -485,17 +485,24 @@ func (a *APIDefinition) IterateSets(iterator dslengine.SetIterator) {
 	}
 	iterator(securitySchemes)
 
-	// And now that we have everything the resources.  The resource
-	// lifecycle handlers dispatch to their children elements, like
-	// Actions, etc..
-	resources := make([]dslengine.Definition, len(a.Resources))
+	// And now that we have everything - the resources. The resource
+	// lifecycle handlers dispatch to their children elements, like Actions,
+	// etc.. We must process parent resources first to ensure that query
+	// string and path parameters are initialized by the time a child
+	// resource action parameters are categorized.
+	resources := make([]*ResourceDefinition, len(a.Resources))
 	i = 0
 	a.IterateResources(func(res *ResourceDefinition) error {
 		resources[i] = res
 		i++
 		return nil
 	})
-	iterator(resources)
+	sort.Sort(byParent(resources))
+	defs := make([]dslengine.Definition, len(resources))
+	for i, r := range resources {
+		defs[i] = r
+	}
+	iterator(defs)
 }
 
 // Reset sets all the API definition fields to their zero value except the default responses and
@@ -890,6 +897,13 @@ func (r *ResourceDefinition) UserTypes() map[string]*UserTypeDefinition {
 	}
 	return types
 }
+
+// byParent makes it possible to sort resources - parents first the children.
+type byParent []*ResourceDefinition
+
+func (p byParent) Len() int           { return len(p) }
+func (p byParent) Swap(i, j int)      { p[i], p[j] = p[j], p[i] }
+func (p byParent) Less(i, j int) bool { return p[j].ParentName == p[i].Name }
 
 // Context returns the generic definition name used in error messages.
 func (cors *CORSDefinition) Context() string {

